From 1385b6d2f6b0cd5ec320d10cee65968baebaf849 Mon Sep 17 00:00:00 2001
From: Yichao Yu <yyc1992@gmail.com>
Date: Sat, 5 Sep 2020 17:36:36 -0400
Subject: [PATCH 8/9] Fix elimination of PiNode for Argument

This causes `PiNode` on `Argument` of type `Argument` to be incorrectly eliminated.
Ironically, this happens for `compact_exprtype` which is exactly where the solution is found.
---
 base/compiler/ssair/ir.jl |  5 +++++
 test/compiler/irpasses.jl | 23 +++++++++++++++++++++++
 2 files changed, 28 insertions(+)

diff --git a/base/compiler/ssair/ir.jl b/base/compiler/ssair/ir.jl
index 4e4cf4a637..bbe6fbab48 100644
--- a/base/compiler/ssair/ir.jl
+++ b/base/compiler/ssair/ir.jl
@@ -981,6 +981,11 @@ function process_node!(compact::IncrementalCompact, result_idx::Int, inst::Instr
                 ssa_rename[idx] = pi_val
                 return result_idx
             end
+        elseif isa(pi_val, Argument)
+            if stmt.typ === compact.ir.argtypes[pi_val.n]
+                ssa_rename[idx] = pi_val
+                return result_idx
+            end
         elseif !isa(pi_val, AnySSAValue) && !isa(pi_val, GlobalRef)
             valtyp = isa(pi_val, QuoteNode) ? typeof(pi_val.value) : typeof(pi_val)
             if valtyp === stmt.typ
diff --git a/test/compiler/irpasses.jl b/test/compiler/irpasses.jl
index 59795d82fe..26bdced759 100644
--- a/test/compiler/irpasses.jl
+++ b/test/compiler/irpasses.jl
@@ -319,3 +319,26 @@ end
 const _some_coeffs = (1,[2],3,4)
 splat_from_globalref(x) = (x, _some_coeffs...,)
 @test splat_from_globalref(0) == (0, 1, [2], 3, 4)
+
+function pi_on_argument(x)
+    if isa(x, Core.Argument)
+        return x.n
+    end
+    return -2
+end
+let code = code_typed(pi_on_argument, Tuple{Any})[1].first.code,
+    nisa = 0, found_pi = false
+    for stmt in code
+        if Meta.isexpr(stmt, :call)
+            callee = stmt.args[1]
+            if (callee === isa || callee === :isa || (isa(callee, GlobalRef) &&
+                                                      callee.name === :isa))
+                nisa += 1
+            end
+        elseif stmt === Core.PiNode(Core.Argument(2), Core.Argument)
+            found_pi = true
+        end
+    end
+    @test nisa == 1
+    @test found_pi
+end
-- 
2.28.0

